googleprojectzero/fuzzilli
# Support disabling mutators through Profile fields

This PR allows mutators to be disabled when the name of the mutator is included in a profile's `disabledMutators` field. Disabling mutators is helpful for:
1. running instances of Fuzzilli with different mutators without having to maintain/build two separate Fuzzilli binaries
2. experimenting with a subset of mutators
3. tailoring Fuzzilli for a target (e.g. JS engines without JIT compilers can disable `JITStressMutator` without maintaining a fork of Fuzzilli or JS engines that have not implemented language features used in `JavaScriptExploreLifting.swift` can disable `ExplorationMutator`)

### Test plan
1. Add print statement to print out list of mutators in `main.swift`
4. Add any mutator into `disabledMutators` field for any profile (e.g. `JITStressMutator`)
5. Run Fuzzilli and check if mutator is present in the list of mutators printed out
```swift
WeightedList<Mutator>(elements: [(elem: Fuzzilli.ExplorationMutator, weight: 3, cumulativeWeight: 0), (elem: Fuzzilli.CodeGenMutator, weight: 2, cumulativeWeight: 3), (elem: Fuzzilli.SpliceMutator, weight: 2, cumulativeWeight: 5), (elem: Fuzzilli.ProbingMutator, weight: 2, cumulativeWeight: 7), (elem: Fuzzilli.InputMutator, weight: 2, cumulativeWeight: 9), (elem: Fuzzilli.InputMutator, weight: 1, cumulativeWeight: 11), (elem: Fuzzilli.OperationMutator, weight: 1, cumulativeWeight: 12), (elem: Fuzzilli.CombineMutator, weight: 1, cumulativeWeight: 13)], totalWeight: 14)
```

### Example usage
```swift
// HermesProfile.swift
let hermesProfile = Profile(
    ...
    disabledMutators: ["JITStressMutator"],
    ...
)
```

```swift
// HermesExperimentalProfile.swift
let hermesProfile = Profile(
    ...
    // Fuzzilli instance focused on variant finding
    disabledMutators: ["JITStressMutator", "CodeGenMutator", "SpliceMutator", "CombineMutator"],
    ...
)
```
